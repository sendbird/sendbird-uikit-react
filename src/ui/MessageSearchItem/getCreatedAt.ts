import type { Locale } from 'date-fns';
import format from 'date-fns/format';
import isToday from 'date-fns/isToday';
import isThisYear from 'date-fns/isThisYear';
import isYesterday from 'date-fns/isYesterday';

interface GetCreatedAtProps {
  createdAt: number;
  locale?: Locale;
  stringSet?: Record<string, string>;
}

// getCreatedAt
export default function ({
  createdAt,
  locale,
  stringSet,
}: GetCreatedAtProps): string {
  const optionalParam = locale ? { locale } : {};
  if (!createdAt) {
    return '';
  }
  if (isToday(createdAt)) {
    return format(createdAt, 'p', optionalParam);
  }
  if (isYesterday(createdAt)) {
    return stringSet?.MESSAGE_STATUS__YESTERDAY || 'Yesterday';
  }
  if (isThisYear(createdAt)) {
    return format(createdAt, 'MMM d', optionalParam);
  }
  return format(createdAt, 'yyyy/M/d', optionalParam);
}
