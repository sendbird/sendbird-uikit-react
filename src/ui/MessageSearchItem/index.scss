@import '../../styles/variables';

.sendbird-message-search-item {
  display: inline-flex;
  flex-direction: row;
  width: 100%;
  height: 80px;
  position: relative;
  cursor: pointer;
  @include themed() {
    border-bottom: 1px solid t(on-bg-4);
    background-color: t(bg-0);
  }

  &:hover {
    @include themed() {
      background-color: t(bg-1);
    }
  }

  .sendbird-message-search-item__left {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 90px;
    height: 80px;
  }
  .sendbird-message-search-item__right {
    display: flex;
    flex-direction: column;
    width: calc(100% - 106px);

    .sendbird-message-search-item__right__sender-name {
      display: flex;
      max-width: 280px;
      height: 16px;
      overflow: hidden;
      text-overflow: ellipsis;
      position: absolute;
      top: 12px;
    }
    .sendbird-message-search-item__right__message-text {
      box-sizing: border-box;
      display: flex;
      width: inherit;
      height: 32px;
      overflow: hidden;
      text-overflow: ellipsis;
      position: absolute;
      top: 36px;
      word-break: break-word;
    }
    .sendbird-message-search-item__right__message-created-at {
      position: absolute;
      top: 12px;
      right: 18px;
    }
  }

  .sendbird-message-search-item__right-footer {
    width: 16px;
  }
}

.sendbird-message-search-item.sendbird-message-search-item--selected {
  @include themed() {
    background-color: t(primary--1-bg--7);
  }
  .sendbird-message-search-item__left {
    @include themed() {
      border-left: 4px solid t(primary-3);
    }
  }
}
