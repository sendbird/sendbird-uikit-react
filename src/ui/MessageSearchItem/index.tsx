import React from 'react';
import './index.scss';
import getCreatedAt from './getCreatedAt';

import Avatar from '../Avatar';
import Label, { LabelTypography, LabelColors } from '../Label';
import { useLocalization } from '../../lib/LocalizationContext';
import { UserMessage } from '@sendbird/chat/message';

interface Props {
  className?: string | Array<string>;
  message: UserMessage;
  selected?: boolean;
  onClick?: (message: UserMessage) => void;
}

export default function MessageSearchItem({
  className,
  message,
  selected,
  onClick,
}: Props): JSX.Element {
  const { createdAt } = message;
  const messageText = message.message;
  // @ts-ignore
  const sender = message.sender || message._sender;
  const { profileUrl, nickname } = sender;
  const { stringSet, dateLocale } = useLocalization();

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-message-search-item',
        selected ? 'sendbird-message-search-item--selected' : '',
      ].join(' ')}
      onClick={(e) => {
        e.stopPropagation();
        onClick?.(message);
      }}
    >
      <div className="sendbird-message-search-item__left">
        <Avatar
          className="sendbird-message-search-item__left__sender-avatar"
          src={profileUrl}
          alt="profile image"
          width="56px"
          height="56px"
        />
      </div>
      <div className="sendbird-message-search-item__right">
        <Label
          className="sendbird-message-search-item__right__sender-name"
          type={LabelTypography.SUBTITLE_2}
          color={LabelColors.ONBACKGROUND_1}
        >
          {nickname || stringSet.NO_NAME}
        </Label>
        <Label
          className="sendbird-message-search-item__right__message-text"
          type={LabelTypography.BODY_2}
          color={LabelColors.ONBACKGROUND_3}
        >
          {messageText}
        </Label>
        <Label
          className="sendbird-message-search-item__right__message-created-at"
          type={LabelTypography.CAPTION_3}
          color={LabelColors.ONBACKGROUND_2}
        >
          {getCreatedAt({ createdAt, locale: dateLocale, stringSet })}
        </Label>
      </div>
      <div className="sendbird-message-search-item__right-footer" />
    </div>
  );
}
