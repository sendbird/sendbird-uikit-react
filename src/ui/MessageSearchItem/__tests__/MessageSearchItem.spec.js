import React from 'react';
import { render } from '@testing-library/react';

import MessageSearchItem from "../index";
import { generateNormalMessage } from '../messageDummyDate.mock';

jest.mock('date-fns/format', () => () => ('mock-date'));


describe('ui/MessageSearchItem', () => {
  it('should render necessary elements', function () {
    const className = 'class-name';
    const { container } = render(
      <MessageSearchItem
        className={className}
        message={generateNormalMessage()}
      />
    );
    expect(
      container.getElementsByClassName(className).length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-message-search-item__left__sender-avatar').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-message-search-item__right__sender-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-message-search-item__right__message-text').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-message-search-item__right__message-created-at').length
    ).toBe(1);
  });

  it('should do a snapshot test of the MessageSearchItem DOM', function () {
    const text = "example-text";
    const { asFragment } = render(
      <MessageSearchItem
        className={text}
        message={generateNormalMessage()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
