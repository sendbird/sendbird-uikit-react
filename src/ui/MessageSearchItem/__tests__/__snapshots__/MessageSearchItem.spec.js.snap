// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/MessageSearchItem should do a snapshot test of the MessageSearchItem DOM 1`] = `
<DocumentFragment>
  <div
    class="example-text sendbird-message-search-item "
  >
    <div
      class="sendbird-message-search-item__left"
    >
      <div
        class="sendbird-message-search-item__left__sender-avatar sendbird-avatar"
        role="button"
        style="height: 56px; width: 56px; z-index: 0; left: 0px;"
        tabindex="0"
      >
        <div
          class="sendbird-avatar-img sendbird-image-renderer"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
        >
          <div
            class="sendbird-image-renderer__image"
            style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_33_512px.png);"
          />
          <img
            alt="profile image"
            class="sendbird-image-renderer__hidden-image-loader"
            src="https://static.sendbird.com/sample/profiles/profile_33_512px.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sendbird-message-search-item__right"
    >
      <span
        class="sendbird-message-search-item__right__sender-name sendbird-label sendbird-label--subtitle-2 sendbird-label--color-onbackground-1"
      >
        Sravan
      </span>
      <span
        class="sendbird-message-search-item__right__message-text sendbird-label sendbird-label--body-2 sendbird-label--color-onbackground-3"
      >
        Hello My name is SendBird.
      </span>
      <span
        class="sendbird-message-search-item__right__message-created-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-2"
      >
        2020/1/23
      </span>
    </div>
    <div
      class="sendbird-message-search-item__right-footer"
    />
  </div>
</DocumentFragment>
`;
